In [2]:
import pandas as pd
import json

In [20]:
rank_file = 'rankings_dict.json'
affiliation_file = 'author_inst_aggregated.csv'

In [21]:
with open(rank_file, 'r') as f:
    rank_dict = json.load(f)

In [6]:
df = pd.read_csv(affiliation_file)

In [59]:
df1 = df[:5000].copy()

In [60]:
df1['rank'] = '1000'

In [61]:
def add_rank(row):
    if row['affiliation'] == '-':
        return row
    split = row['affiliation'].split(',')
    country = split[-1].strip()
    uni = ','.join(split[:-1])
    if country in rank_dict and uni in rank_dict[country]:
        row['rank'] = rank_dict[country][uni]
    return row

In [62]:
df1.apply(add_rank, axis=1)

pid              affiliation  rank
0      31/4529                        -  1000
1      31/7737                        -  1000
2      31/1499                        -  1000
3     31/10325                        -  1000
4      31/3692  Nokia Research, Finland  1000
...        ...                      ...   ...
4995   31/3147                        -  1000
4996  31/11446                        -  1000
4997   31/5775                        -  1000
4998   31/8357                        -  1000
4999   31/1264                        -  1000

[5000 rows x 3 columns]

In [63]:
df1['rank'].unique()

array(['1000', '575', '2', '6', '3', '275', '325', '125', '44', '175',
       '425', '24', '525', '375', '35', '225', '475', '75', '29', '14',
       '34', '13', '19', '5'], dtype=object)

In [64]:
df1['rank'].value_counts()

1000    4943
325        9
275        7
125        4
2          4
575        3
375        3
3          3
175        3
44         2
425        2
35         2
475        2
75         2
225        2
525        1
24         1
19         1
5          1
29         1
13         1
6          1
14         1
34         1
Name: rank, dtype: int64

In [65]:
df1.affiliation.value_counts()

-                                                         4771
affiliation                                                  9
Stanford University, USA                                     4
Carnegie Mellon University, USA                              3
University of Wisconsin, USA                                 3
                                                          ... 
University of So Paulo, Department of Computer Science       1
Portuguese Military Academy, Lisbon, Portugal                1
University of Nevada, USA                                    1
Guangzhou University, China                                  1
University of Cantabria, Spain                               1
Name: affiliation, Length: 195, dtype: int64

In [77]:
import cProfile

In [74]:
def time_test(num):
    df1 = df[:num].copy()
    df1['rank'] = '1000'
    df1.apply(add_rank, axis=1)

In [81]:
cProfile.run('time_test(100000)')

         42043097 function calls (41638761 primitive calls) in 28.221 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.000    0.000    0.000    0.000 <__array_function__ internals>:2(append)
        2    0.000    0.000    0.000    0.000 <__array_function__ internals>:2(argsort)
        1    0.000    0.000    0.000    0.000 <__array_function__ internals>:2(array_equal)
        3    0.000    0.000    0.000    0.000 <__array_function__ internals>:2(atleast_2d)
        1    0.000    0.000    0.000    0.000 <__array_function__ internals>:2(bincount)
        1    0.000    0.000    0.000    0.000 <__array_function__ internals>:2(broadcast_to)
        8    0.000    0.000    0.006    0.001 <__array_function__ internals>:2(concatenate)
        7    0.000    0.000    0.000    0.000 <__array_function__ internals>:2(copyto)
        2    0.000    0.000    0.000    0.000 <__array_function__ internals>:2(ndim)
        2    0.

   100006    0.336    0.000    5.004    0.000 managers.py:1467(__init__)
   913783    0.225    0.000    0.225    0.000 managers.py:1520(_block)
        1    0.000    0.000    0.000    0.000 managers.py:1544(index)
   604232    0.408    0.000    0.681    0.000 managers.py:1548(dtype)
        5    0.000    0.000    0.000    0.000 managers.py:1562(external_values)
   101087    0.061    0.000    0.096    0.000 managers.py:1565(internal_values)
   104229    0.132    0.000    0.358    0.000 managers.py:1568(get_values)
        1    0.000    0.000    0.000    0.000 managers.py:1582(_consolidate_inplace)
       20    0.000    0.000    0.000    0.000 managers.py:163(shape)
        1    0.000    0.000    0.007    0.007 managers.py:1643(create_block_manager_from_blocks)
       60    0.000    0.000    0.000    0.000 managers.py:165(<genexpr>)
        1    0.021    0.021    4.650    4.650 managers.py:1667(create_block_manager_from_arrays)
       18    0.000    0.000    0.000    0.000 managers.py:16

       15    0.000    0.000    0.000    0.000 {pandas._libs.lib.is_bool}
   204230    0.040    0.000    0.040    0.000 {pandas._libs.lib.is_float}
        5    0.000    0.000    0.000    0.000 {pandas._libs.lib.is_integer}
   300026    0.302    0.000    0.592    0.000 {pandas._libs.lib.is_list_like}
   204232    0.042    0.000    0.042    0.000 {pandas._libs.lib.is_scalar}
        1    0.000    0.000    0.000    0.000 {pandas._libs.lib.item_from_zerodim}
        1    0.000    0.000    0.000    0.000 {pandas._libs.lib.map_infer}
        6    0.001    0.000    0.001    0.000 {pandas._libs.lib.maybe_convert_objects}
   208459    0.303    0.000    0.731    0.000 {pandas._libs.lib.values_from_object}
        1    0.004    0.004    0.005    0.005 {pandas._libs.reduction.compute_reduction}




num rows | time in seconds

100: 0.06

1000: 0.310

10000: 3.139

100000: 28.221

expected for 2.4million rows: ~672 seconds = 11 mins+

In [84]:
df['rank'] = '1000'
df.apply(add_rank, axis=1)

pid              affiliation  rank
0         31/4529                        -  1000
1         31/7737                        -  1000
2         31/1499                        -  1000
3        31/10325                        -  1000
4         31/3692  Nokia Research, Finland  1000
...           ...                      ...   ...
2390060  16/10858                        -  1000
2390061   16/4994                        -  1000
2390062   16/6180                        -  1000
2390063   16/8515                        -  1000
2390064   16/8852                        -  1000

[2390065 rows x 3 columns]

In [86]:
df.to_csv('author_aff_rank.csv', index=False)